### PR TITLE
[18.09] Add native rpm compilation for devicemapper

### PR DESCRIPTION
**Pull request by seemethere**
devicemapper cannot be enabled on a statically compiled binary so we
dynamically compile it on a rhel based distribution in order to enable
devicemapper usage.

There will also need to be a forward port of this patch to master

Signed-off-by: Eli Uriegas <eli.uriegas@docker.com>

### CODE REVIEW

File: image/Dockerfile.engine-dm [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+RUN yum install -y \
+    bash \
+    ca-certificates \
+    cmake \
+    gcc \
+    git \
+    glibc-static \
+    libselinux-devel \
+    libgcc-6-dev \
+    libltdl-dev \
```
Comment (kolyshkin) [inline image/Dockerfile.engine-dm]: I am pretty sure there are no such packages in centos.

Comment (andrewhsu) [reply image/Dockerfile.engine-dm]: i think what's needed for centos is this:
```
make
cmake
gcc
git
glibc-static
libtool-ltdl-devel
libseccomp-devel
device-mapper-devel
btrfs-progs-devel
```

Comment (seemethere) [reply image/Dockerfile.engine-dm]: Removing:
```diff
-    libselinux-devel \
-    libgcc-6-dev \
-    libltdl-dev \
```

Comment (kolyshkin) [reply image/Dockerfile.engine-dm]: also, glibc-static is not needed as we're doing a dynamic build

File: rpm/Makefile [hunk 2]
```diff
@@ -128,9 +128,9 @@ $(DOCKER2OCI):
 
 # offline bundle
 rpmbuild/SOURCES/engine.tar: $(DOCKER2OCI)
-	$(MAKE) -C ../image ENGINE_IMAGE=$(ENGINE_IMAGE) image-linux
+	$(MAKE) -C ../image ENGINE_IMAGE=$(ENGINE_IMAGE) image-linux-dm
```
[the remark this answers is not in the document]
Comment (seemethere) [reply rpm/Makefile]: Added in 358b54d

File: image/Makefile [hunk 3]
```diff
@@ -18,34 +29,30 @@ help: ## show make targets
...
+	echo $(IMAGE_WITH_TAG) > $@
 
 engine-$(ARCH).tar: image-linux
 	docker save -o $@ $$(cat $<)
 
-
 .PHONY: release
 release:
-	docker push $(DOCKER_HUB_ORG)/$(ENGINE_IMAGE):$(STATIC_VERSION).$(ARCH)
+	docker push $(IMAGE_WITH_TAG)
```
Comment (jose-bigio) [inline image/Makefile]: Does the release target need to account for whether the image name has a `-dm` appended to the end of it?

File: image/Makefile [hunk 4]
```diff
@@ -18,34 +29,30 @@ help: ## show make targets
...
 clean: ## remove build artifacts
 	-$(RM) $(ENGINE_DIR)/Dockerfile.engine
-	-docker rmi $(DOCKER_HUB_ORG)/$(ENGINE_IMAGE):$(STATIC_VERSION)
-	-rm -f image-linux
+	-$(RM) $(ENGINE_DIR)/Dockerfile.engine-dm
+	-docker rmi $(IMAGE_WITH_TAG)
+	-rm -f image-linux-*
 
 .PHONY: image
 image: image-linux
```
Comment (dhiltgen) [inline image/Makefile]: Should this build both images so we make sure they're "tied at the hip" and always built together?  

Also related, I think there are other places in the CI pipeline where it builds "image-linux" which means it's not going to build this image the way this is structured, which seems potentially wrong.

File: image/Makefile [hunk 5]
```diff
@@ -18,34 +29,30 @@ help: ## show make targets
...
 
-
-$(ENGINE_DIR)/Dockerfile.engine:
-	cp Dockerfile.engine $(ENGINE_DIR)
+$(ENGINE_DIR)/Dockerfile.%: Dockerfile.%
+	cp $< $@
 
 # builds across multiple archs because the base images
 # utilize manifests
 image-linux: $(ENGINE_DIR)/Dockerfile.engine
```
Comment (dhiltgen) [inline image/Makefile]: As an alternative, maybe this target could be restructured to build both by defining a new target for the overlay image and one for dm?
```
image-linux: image-linux-o2 image-linux-dm

image-linux-o2: $(ENGINE_DIR)/Dockerfile.engine
        ...
```

Comment (dhiltgen) [reply image/Makefile]: Hmm... now that I'm looking I can't seem to find where I saw it, but I swear somewhere in the CI pipeline scripts I saw a "make image-linux" (or something along those lines)

**Comment by seemethere** [comment]
Closing this, will re-open in a new PR

**Pull request closed without merging**